crates/sui-transactional-test-runner/src/lib.rs

```diff
@@ -9,11 +9,154 @@ pub mod test_adapter;
+
 #[cfg_attr(not(msim), tokio::main)]
 #[cfg_attr(msim, msim::main)]
 pub async fn run_test(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
     run_test_impl::<SuiTestAdapter>(path, Some(&*PRE_COMPILED)).await?;
     Ok(())
 }
+
+pub struct ValidatorWithFullnode {
+    pub validator: Arc<AuthorityState>,
+    pub fullnode: Arc<AuthorityState>,
+    pub kv_store: Arc<TransactionKeyValueStore>,
+}
+
+#[allow(unused_variables)]
+/// TODO: better name?
+#[async_trait::async_trait]
+pub trait TransactionalAdapterTestable: Send + Sync + ObjectStore {
+    async fn execute_txn(
+        &mut self,
+        transaction: Transaction,
+    ) -> anyhow::Result<(TransactionEffects, Option<ExecutionError>)>;
+
+    async fn create_checkpoint(&mut self) -> anyhow::Result<VerifiedCheckpoint> {
+        unimplemented!("create_checkpoint not supported")
+    }
+
+    async fn advance_clock(
+        &mut self,
+        duration: std::time::Duration,
+    ) -> anyhow::Result<TransactionEffects> {
+        unimplemented!("advance_clock not supported")
+    }
+
+    async fn advance_epoch(&mut self) -> anyhow::Result<()> {
+        unimplemented!("advance_epoch not supported")
+    }
+
+    async fn request_gas(
+        &mut self,
+        address: SuiAddress,
+        amount: u64,
+    ) -> anyhow::Result<TransactionEffects> {
+        unimplemented!("request_gas not supported")
+    }
+
+    async fn dev_inspect_transaction_block(
+        &self,
+        sender: SuiAddress,
+        transaction_kind: TransactionKind,
+        gas_price: Option<u64>,
+    ) -> SuiResult<DevInspectResults> {
+        unimplemented!("dev_inspect_transaction_block not supported")
+    }
+
+    async fn query_events(
+        &self,
+        query: EventFilter,
+        // If `Some`, the query will start from the next item after the specified cursor
+        cursor: Option<EventID>,
+        limit: usize,
+        descending: bool,
+    ) -> SuiResult<Vec<SuiEvent>> {
+        unimplemented!("query_events not supported")
+    }
+}
+
+#[async_trait::async_trait]
+impl TransactionalAdapterTestable for ValidatorWithFullnode {
+    async fn execute_txn(
+        &mut self,
+        transaction: Transaction,
+    ) -> anyhow::Result<(TransactionEffects, Option<ExecutionError>)> {
+        let with_shared = transaction
+            .data()
+            .intent_message()
+            .value
+            .contains_shared_object();
+        let (_, effects, execution_error) = send_and_confirm_transaction_with_execution_error(
+            &self.validator,
+            Some(&self.fullnode),
+            transaction,
+            with_shared,
+        )
+        .await?;
+        Ok((effects.into_data(), execution_error))
+    }
+
+    async fn dev_inspect_transaction_block(
+        &self,
+        sender: SuiAddress,
+        transaction_kind: TransactionKind,
+        gas_price: Option<u64>,
+    ) -> SuiResult<DevInspectResults> {
+        self.fullnode
```

Comment:
generally we want to fetch data from validator unless only possible/ easier through fullnode?

Reply:
i see, this is just mirroring the existing impl. on test_adapter